tests/integration/test_concatenate_stack.py

```diff
@@ -93,24 +99,341 @@ def test_concatenate(size, num, a):
+    b = [[5, 6]]
+    axis = 0
+    out_np = np.zeros((3, 2))
+    out_num = num.array(out_np)
+
+    np.concatenate((np.array(a), np.array(b)), axis=axis, out=out_np)
+    num.concatenate((num.array(a), num.array(b)), axis=axis, out=out_num)
+    assert np.array_equal(out_np, out_num)
+
+
+@pytest.mark.parametrize(
+    "dtype", (np.float32, np.int32), ids=lambda dtype: f"(dtype={dtype})"
+)
+def test_concatenate_dtype(dtype):
+    a = [[1, 2], [3, 4]]
+    b = [[5, 6]]
+    axis = 0
+
+    res_np = np.concatenate((np.array(a), np.array(b)), axis=axis, dtype=dtype)
+    res_num = num.concatenate(
+        (num.array(a), num.array(b)), axis=axis, dtype=dtype
+    )
+    assert np.array_equal(res_np, res_num)
+
+
+@pytest.mark.parametrize(
+    "casting",
+    ("no", "equiv", "safe", "same_kind", "unsafe"),
+    ids=lambda casting: f"(casting={casting})",
+)
+def test_concatenate_casting(casting):
+    a = [[1, 2], [3, 4]]
+    b = [[5, 6]]
+    axis = 0
+
+    res_np = np.concatenate(
+        (np.array(a), np.array(b)), axis=axis, casting=casting
+    )
+    res_num = num.concatenate(
+        (num.array(a), num.array(b)), axis=axis, casting=casting
+    )
+    assert np.array_equal(res_np, res_num)
+
+
+class TestConcatenateErrors:
+    def test_zero_arrays(self):
+        expected_exc = ValueError
+        arrays = ()
+        axis = None
+        with pytest.raises(expected_exc):
+            np.concatenate(arrays, axis=axis)
+        with pytest.raises(expected_exc):
+            num.concatenate(arrays, axis=axis)
+
+    @pytest.mark.parametrize(
+        "arrays",
+        (
+            pytest.param((1,), marks=pytest.mark.xfail),
+            pytest.param((1, 2), marks=pytest.mark.xfail),
+            (1, [3, 4]),
+        ),
+        ids=lambda arrays: f"(arrays={arrays})",
+    )
+    def test_scalar_axis_is_not_none(self, arrays):
+        # For (1,) and (1, 2),
+        # In Numpy, it raises ValueError
+        # In cuNumeric, it raises IndexError
+        expected_exc = ValueError
+        axis = 0
+        with pytest.raises(expected_exc):
+            np.concatenate(arrays, axis=axis)
+        with pytest.raises(expected_exc):
+            num.concatenate(arrays, axis=axis)
+
+    @pytest.mark.parametrize(
+        "arrays",
+        (
+            ([[1, 2], [3, 4]], [5, 6]),
+            ([[1, 2], [3, 4], [5, 6]], [[7, 8], [9, 10]]),
+            pytest.param(
+                ([[1, 2], [3, 4]], [[5, 6]]), marks=pytest.mark.xfail
+            ),
+        ),
+        ids=lambda arrays: f"(arrays={arrays})",
+    )
+    def test_arrays_mismatched_shape(self, arrays):
+        # for ([[1, 2], [3, 4]], [[5, 6]]),
+        # In Numpy, it raises ValueError
+        # In cuNumeric, it pass
+        expected_exc = ValueError
+        axis = 1
+        with pytest.raises(expected_exc):
+            np.concatenate(arrays, axis=axis)
+        with pytest.raises(expected_exc):
+            num.concatenate(arrays, axis=axis)
```

Comment:
I think the check for the number of dimensions matching in `concatenate` is not strict enough. @robinw0928 can you try adjusting as follows:
```
--- a/cunumeric/module.py
+++ b/cunumeric/module.py
@@ -1461,8 +1461,8 @@ def check_shape_dtype(
             "must have the same number of dimensions"
         )
     if ndim > 1 and _builtin_any(
-        shape[1:axis] != inp.shape[1:axis]
-        and shape[axis + 1 :] != inp.shape[axis + 1 :]
+        shape[:axis] != inp.shape[:axis]
+        or shape[axis + 1 :] != inp.shape[axis + 1 :]
         for inp in inputs
     ):
         raise ValueError(
```

Reply:
@manopapad You are right. Actually, I found we may need to change 6 places of source code for concatenate and stack.
List:
1. In check_shape_dtype, and should be replaced by or. Note: shape[1:axis]!=inp.shape[1:axis] returns a single value: True or False.
2. In check_shape_dtype, change shape[1:axis] to shape[0:axis]. I think the first size of shape should also be checked. Or is it omitted purposefully?
3. In check_shape_dtype, check shapes of all input arrays for stack. This is necessary for stack.
4. In check_shape_dtype, change ndim > 1 to ndim >= 1. This is necessary for checking 1-D array size for column_stack/concatenate.
5. In check_shape_dtype, normalize axis if axis < 0 and axis >= -ndim. Otherwise, axis=-1 would cause bugs in shape[axis+1:] for append API.
6. add normalize_axis_index in def stack. This is necessary for the case where axis < 0 and axis >= -(ndim+1)

May I submit a new PR to address theses issues? In this PR, we only focus on test code.

Reply:
Yes, you're right, we should leave all that for another PR.